crates/bevy_animation/src/lib.rs

```diff
@@ -479,174 +485,217 @@ impl AnimationPlayer {
         }
+        dirty_entities.insert(entity);
+    }
+
+    // Find the roots of the trees in the forest.
+    worklist = dirty_entities
+        .iter()
+        .filter(|&&entity| {
+            // Do we have a clean parent? If so, we're a root.
+            !candidate_targets
+                .get(entity)
+                .is_ok_and(|(_, parent, _, _, _)| {
+                    parent.is_some_and(|parent| dirty_entities.contains(&parent.get()))
+                })
+        })
+        .cloned()
+        .collect();
+
+    // Compute new `AnimationTarget` components.
+    let mut new_targets: EntityHashMap<Entity, AnimationTarget> = EntityHashMap::default();
+    let mut removed_targets: EntityHashSet<Entity> = EntityHashSet::default();
+    while let Some(entity) = worklist.pop() {
+        let (_, parent, kids, name, _) = candidate_targets
+            .get(entity)
+            .expect("Shouldn't have dirtied an entity that wasn't a target");
+
+        // Determine the path based on our parent.
+        let mut new_target = AnimationTarget::new();
         if let Some(parent) = parent {
-            current = parent.get();
+            // Figure out the parent path. First, if we've already computed the path
+            // of the parent, just use that.
+            if let Some(parent_target) = new_targets.get(&parent.get()) {
+                new_target = parent_target.clone();
+            }
+
+            // Otherwise, if this is a root of a tree in the forest, we know the
+            // parent's `AnimationTarget` is up to date, so fetch that.
+            if new_target.path.is_empty() {
+                if let Ok((_, _, _, _, Some(parent_target))) = candidate_targets.get(parent.get()) {
+                    new_target = parent_target.clone();
+                }
+            }
+
+            // If we still didn't find an `AnimationTarget` for the parent, we could
+            // be an immediate child of an animation player. Handle that case.
+            if new_target.path.is_empty() {
+                if let Ok((_, _, _, name)) = players.get(parent.get()) {
+                    new_target = AnimationTarget {
+                        root: parent.get(),
+                        path: EntityPath::from(name.clone()),
+                    };
+                }
+            }
+        }
+
+        // If we still don't have an animation target after all that, we're part
+        // of a subtree that was once an animation target but got disconnected.
+        // Remove the `AnimationTarget` component entirely.
+        if new_target.path.is_empty() {
+            removed_targets.insert(entity);
         } else {
-            return true;
+            // Push our own name onto the path and record it.
+            new_target.path.parts.push(name.clone());
+            new_targets.insert(entity, new_target);
+        }
+
+        // Enqueue children.
+        if let Some(kids) = kids {
+            worklist.extend(kids);
         }
```

Comment:
Do we need to use the same loop for children? Can't we simplify their handling quite a bit, since we know they're chilren of an actual (non-removed) `AnimationTarget`, and on top of that direct children, so their path should be almost identical, no?